onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc

```diff
@@ -2274,193 +2448,188 @@ common::Status TensorrtExecutionProvider::Compile(const std::vector<FusedNodeAnd
+            LOGS_DEFAULT(WARNING) << "[TensorRT EP] Builder optimization level can only be used on TRT 8.6 onwards!";
+          }
+          if (trt_state->auxiliary_streams >= 0) {
+            LOGS_DEFAULT(WARNING) << "[TensorRT EP] Auxiliary streams can only be set on TRT 8.6 onwards!";
+          }
 #endif
-        // limit used tactic sources
-        if (trt_state->filter_tactic_sources) {
-          nvinfer1::TacticSources tactics = trt_config->getTacticSources();
-          tactics |= trt_state->tactic_sources;
-          trt_config->setTacticSources(tactics);
-          LOGS_DEFAULT(VERBOSE) << "[TensorRT EP] Tactic sources are limited using bitmask " << tactics;
-        }
-
-        // Load timing cache from file. Create a fresh cache if the file doesn't exist
-        std::unique_ptr<nvinfer1::ITimingCache> timing_cache = nullptr;
-        if (trt_state->timing_cache_enable) {
-          std::vector<char> loaded_timing_cache = loadTimingCacheFile(timing_cache_path);
-          timing_cache.reset(trt_config->createTimingCache(static_cast<const void*>(loaded_timing_cache.data()), loaded_timing_cache.size()));
-          if (timing_cache == nullptr) {
-            return ORT_MAKE_STATUS(ONNXRUNTIME, EP_FAIL,
-                                   "TensorRT EP could not create timing cache: " + timing_cache_path);
+          // limit used tactic sources
+          if (trt_state->filter_tactic_sources) {
+            nvinfer1::TacticSources tactics = trt_config->getTacticSources();
+            tactics |= trt_state->tactic_sources;
+            trt_config->setTacticSources(tactics);
+            LOGS_DEFAULT(VERBOSE) << "[TensorRT EP] Tactic sources are limited using bitmask " << tactics;
           }
-          trt_config->setTimingCache(*timing_cache, force_timing_cache_match_);
-          if (detailed_build_log_) {
-            LOGS_DEFAULT(VERBOSE) << "[TensorRT EP] Deserialized timing cache from " + timing_cache_path;
+
+          // Load timing cache from file. Create a fresh cache if the file doesn't exist
+          std::unique_ptr<nvinfer1::ITimingCache> timing_cache = nullptr;
+          if (trt_state->timing_cache_enable) {
+            std::vector<char> loaded_timing_cache = loadTimingCacheFile(timing_cache_path);
+            timing_cache.reset(trt_config->createTimingCache(static_cast<const void*>(loaded_timing_cache.data()), loaded_timing_cache.size()));
+            if (timing_cache == nullptr) {
+              return ORT_MAKE_STATUS(ONNXRUNTIME, EP_FAIL,
+                                     "TensorRT EP could not create timing cache: " + timing_cache_path);
+            }
+            trt_config->setTimingCache(*timing_cache, force_timing_cache_match_);
+            if (detailed_build_log_) {
+              LOGS_DEFAULT(VERBOSE) << "[TensorRT EP] Deserialized timing cache from " + timing_cache_path;
+            }
           }
-        }
 
-        // Build engine
-        {
-          auto lock = GetApiLock();
```

Comment:
should this call to GetApiLock() be removed?

Reply:
I think this lock is unnecessary.
The current TRT EP already has a [lock](https://github.com/microsoft/onnxruntime/blob/rel-1.15.1/onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc#L2214) for the whole compute function  and this "Build engine" block is already covered.
For this PR, this "Build engine" block also covered by the lock [here](https://github.com/microsoft/onnxruntime/blob/chi/one_trt_context_per_thread/onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc#L2464).

Reply:
Discussed offline. We should add back api lock which prevents multiple engines from being built at the same time, even for different subgraphs. (Even though nvidia's doc didn't say you can't have multiple engine builds at the same time, we should avoid this due to it might mess up the timings of their tactics selection/profiling.)